File: cs/Minsk/CodeAnalysis/EvaluationResult.cs
using System.Collections.Immutable;

namespace Minsk.CodeAnalysis;

public sealed class EvaluationResult
{
    public readonly ImmutableArray<Diagnostic> Diagnostics;
    public readonly object? Value;

    internal EvaluationResult(ImmutableArray<Diagnostic> diagnostics, object? value)
    {
        Diagnostics = diagnostics;
        Value = value;
    }
}

File: nim/src/minsk/codeAnalysis/evaluationResult.nim
import minsk/codeAnalysis/diagnostic
import minsk/minskObject

type
  EvaluationResult* = object
    case success*: bool
    of true:
      value*: MinskObject
    of false:
      diagnostics*: seq[Diagnostic]

proc evaluationResultOk*(value: MinskObject): EvaluationResult =
  EvaluationResult(
    success: true,
    value: value,
  )

proc evaluationResultError*(diagnostics: seq[Diagnostic]): EvaluationResult =
  EvaluationResult(
    success: false,
    diagnostics: diagnostics,
  )
